File: problems/1402_reducing_dishes/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int maxSatisfaction(vector<int>& satisfaction) {
        sort(satisfaction.rbegin(), satisfaction.rend());
        int res = 0, prefix = 0;
        for (int s : satisfaction) {
            prefix += s;
            if (prefix <= 0) break;
            res += prefix;
        }
        return res;
    }
};

File: problems/1402_reducing_dishes/solution.py
from typing import List


class Solution:
    def maxSatisfaction(self, satisfaction: List[int]) -> int:
        satisfaction.sort(reverse=True)
        res = 0
        prefix = 0
        for s in satisfaction:
            prefix += s
            if prefix <= 0:
                break
            res += prefix
        return res
